base/applications/taskmgr/perfdata.c

```diff
@@ -572,21 +572,48 @@ BOOL PerfDataGetCommandLine(ULONG Index, LPWSTR lpCommandLine, ULONG nMaxCount)
     /* [A] Search for a string already in cache? If so, use it. If not, check if invalid and free it if so. */
     while (cache && cache->pnext != NULL)
     {
+        EnterCriticalSection(&PerfDataCriticalSection);
+
+        /* Because threading *might* free the cache we're reading from, check if we still exist. */
+        if (cache == NULL)
+        {
+            if (old_cache != NULL)
+            {
+                /* Use the previous cache that wasnt deallocated. */
+                cache = old_cache;
+            }
+            else if (global_cache != NULL)
+            {
+                /* Start at the beginning, because we have no idea where we are. */
+                cache = global_cache;
+            }
+            else
+            {
+                /* Everything is screwed, lets pretend that this never happened. */
+                break;
```

Comment:
Isn't `LeaveCriticalSection(&PerfDataCriticalSection);` needed here before break?

Reply:
Yes, yes it is.